### ISSUE TITLE
Refactor kubernetes scheduler and controller manager to enable code reuse

### ISSUE THREAD

**Issue by irfanurrehman** [kubernetes/federation#42]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Apr 29, 2016 at 20:06 GMT_
_Originally opened as https://github.com/kubernetes/kubernetes/issues/24996_

----

Forked from https://github.com/kubernetes/kubernetes/pull/24038#issuecomment-215610094

We are adding federation scheduler and controller manager as per the plan in https://github.com/kubernetes/kubernetes/issues/23653.

There is obviously some common code that is shared with kubernetes controller manager and kubernetes scheduler.
We want to pull out that code into libraries to enable code reuse rather than having to duplicate code.

We need to figure out the right name and directory structure for such libraries.

cc @davidopp @madhusudancs @kubernetes/goog-control-plane @kubernetes/sig-cluster-federation

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Apr 29, 2016 at 20:06 GMT_

----

Note that there was same issue with kubernetes and federation apiserver. We now have a library `pkg/genericapiserver` that solves that problem.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Apr 29, 2016 at 20:10 GMT_

----

Marking this 1.3 since this is also required for cluster controller which is planned for 1.3.
Scheduler is not strictly required for 1.3

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Apr 29, 2016 at 20:52 GMT_

----

For reference, cluster controller PR is https://github.com/kubernetes/kubernetes/pull/23430 and federation scheduler PR is https://github.com/kubernetes/kubernetes/pull/24038

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Monday May 02, 2016 at 23:45 GMT_

----

I discussed this with @davidopp and summarizing our discussion here:

controller-manager:
- Common code: Setting up the binary: command line arguments (adding flags, parsing them, setting defaults, validating values). Mostly the code in `cmd/` directory.
- Different code: Which controllers to run, business logic in each controller.

We do want to refactor common code, but dont necessarily want to block the controller PR on that.

Scheduler: There should not be a lot of common code between federation and kubernetes schedulers.
- The core business logic in both the schedulers will be very different (kube scheduler watches pods and sends POST requests to pod bindings while fed scheduler will watch replicasets and POST sub replicasets).
- There will be some common command line flags related to talking to API server (URL, auth, etc) but that is same for any component talking to API server (like kubelet). nothing special with federation scheduler

So we can start with federated scheduler independently and if it later turns out that there is a lot of common code (which we hope will not happen), we can refactor.

So neither scheduler nor controller manager is blocked on refactoring.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/davidopp"><img src="https://avatars2.githubusercontent.com/u/8264113?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [davidopp](https://github.com/davidopp)**
_Tuesday May 03, 2016 at 00:16 GMT_

----

> We do want to refactor common code, but dont necessarily want to block the controller PR on that.

Yeah, it sounds like we don't have time to figure out and implement exactly the "right" way to do the refactoring. But we need to do it right after 1.3 ships. We can't leave this duplicated code around.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Monday Jul 25, 2016 at 22:48 GMT_

----

This fits into what @mwielgus is working on.  Tentatively assigning to him so that we don't lose track of it.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/dims"><img src="https://avatars2.githubusercontent.com/u/23304?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [dims](https://github.com/dims)**
_Tuesday Nov 15, 2016 at 16:22 GMT_

----

switching this to 1.6 as it's too late for 1.5. ok? (please switch it right back if you disagree)

**Comment by irfanurrehman** [comment]
<a href="https://github.com/timothysc"><img src="https://avatars0.githubusercontent.com/u/169553?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [timothysc](https://github.com/timothysc)**
_Thursday Dec 08, 2016 at 00:21 GMT_

----

@jayunit100 This is part of the 1.6 cleanup items for you.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/ethernetdan"><img src="https://avatars1.githubusercontent.com/u/4014393?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ethernetdan](https://github.com/ethernetdan)**
_Monday Mar 13, 2017 at 22:35 GMT_

----

Moving to 1.7 as late to happen in 1.6. Feel free to switch back if this is incorrect.

**Comment by irfanurrehman** [comment]
cc @nikhiljindal

**Comment by shashidharatd** [comment]
This issue is no longer relevant. There is no scheduler code in federation anymore.
/close
